File: fastapi/plugs/websocket/ConnectionManager.py
from business.services.loggerInteface import ILogger
from fastapi import WebSocket
import json

class ConnectionModel:
    board_id: str
    socket: WebSocket
    def __init__(self, board_id: str, socket: WebSocket):
        self.board_id = board_id
        self.socket = socket

class ConnectionManager:
    logger: ILogger
    def __init__(self, logger: ILogger):
        self.active_connections: list[ConnectionModel] = []
        self.logger = logger

    async def connect(self, board_id: str, websocket: WebSocket):
        await websocket.accept()
        self.active_connections.append(ConnectionModel(board_id, websocket))
        self._connection_count()
        await self.update_participations_count(board_id)

    async def update_participations_count(self, board_id: str):
        all_connections_same_board = list(filter(lambda x: x.board_id == board_id, self.active_connections))
        message_object = {
            "type": "PARTICIPANTS_COUNT_UPDATE",
            "data": len(all_connections_same_board)
        }
        json_string = json.dumps(message_object)
        for connection in all_connections_same_board:
            await connection.socket.send_text(json_string)

    def disconnect(self, websocket: WebSocket):
        self.active_connections = list(filter(lambda x: x.socket != websocket, self.active_connections))
        self._connection_count()

    async def send_personal_message(self, message: str, websocket: WebSocket):
        await websocket.send_text(message)

    async def broadcast(self, message: str):
        for connection in self.active_connections:
            await connection.socket.send_text(message)

    async def to_board_users(self, board_id: str, message: str, currentUser: WebSocket):
        for connection in filter(lambda x: x.board_id == board_id and x.socket != currentUser, self.active_connections):
            await connection.socket.send_text(message)
    def _connection_count(self):
        self.logger.log(f"{len(self.active_connections)} connections left!")

File: reactive-UI/src/dependencies/socket-communication/connection.manager.ts
import WebSocket from 'ws';

export interface ILogger {
    log(message: string): void;
}

export class ConsoleLogger implements ILogger {
    log(message: string): void {
        console.log(message);
    }
}

class ConnectionModel {
  boardId: string;
  socket: WebSocket;

  constructor(boardId: string, socket: WebSocket) {
    this.boardId = boardId;
    this.socket = socket;
  }
}

export class ConnectionManager {
  private activeConnections: ConnectionModel[] = [];
  private logger: ILogger;

  constructor(logger: ILogger) {
    this.logger = logger;
  }

  async connect(boardId: string, socket: WebSocket): Promise<void> {
    this.activeConnections.push(new ConnectionModel(boardId, socket));
    this._connectionCount();
    await this.updateParticipationsCount(boardId);
  }

  async updateParticipationsCount(boardId: string): Promise<void> {
    const allConnectionsSameBoard = this.activeConnections.filter(connection => connection.boardId === boardId);
    const messageObject = {
      type: 'PARTICIPANTS_COUNT_UPDATE',
      data: allConnectionsSameBoard.length
    };
    const jsonString = JSON.stringify(messageObject);
    for (const connection of allConnectionsSameBoard) {
      connection.socket.send(jsonString);
    }
  }

  disconnect(socket: WebSocket): void {
    this.activeConnections = this.activeConnections.filter(connection => connection.socket !== socket);
    this._connectionCount();
  }

  async sendPersonalMessage(message: string, socket: WebSocket): Promise<void> {
    socket.send(message);
  }

  async broadcast(message: string): Promise<void> {
    for (const connection of this.activeConnections) {
      connection.socket.send(message);
    }
  }

  async toBoardUsers(boardId: string, message: string, currentUser: WebSocket): Promise<void> {
    for (const connection of this.activeConnections.filter(connection => connection.boardId === boardId && connection.socket !== currentUser)) {
      connection.socket.send(message);
    }
  }

  private _connectionCount(): void {
    this.logger.log(`${this.activeConnections.length} connections left!`);
  }
}
